guide/src/async-await.md

```diff
@@ -0,0 +1,240 @@
+# Async / Await
+
+Both Python and Rust have support for async functions, but bindings for these functions are not as 
+straightforward as they are for blocking functions. Both languages have a fairly distinct model for 
+async functions and Python's needs in particular can sometimes be restrictive.
+
+[pyo3-asyncio](https://github.com/awestlake87/pyo3-asyncio) was created to provide conversions 
+between async Python and async Rust as well as manage the nitty gritty details of Python's event 
+loop.
+
+## Awaiting an Async Python Function in Rust
+
+Let's take a look at a dead simple async Python function:
+
+```python
+async def py_sleep():
+    await asyncio.sleep(1)
+```
+
+This function simply sleeps for 1 second and returns. So what does this look like to PyO3? 
+
+First, it helps to have a little background on Python's async functions. 
+Async functions in python are simply functions that return a `coroutine` object. You can read more about 
+`coroutine` objects in the [Python 3 docs](https://docs.python.org/3/library/asyncio-task.html), but
+for our purposes, we really don't need to know much about them. The key factor here is that calling
+an `async` function is _just like calling a regular function_, the only difference is that we have
+to do something special with the object that it returns.
+
+Normally in Python, that something special is the `await` keyword, but in Rust, we don't have the 
+luxury of using Python's syntax. Luckily, Rust also has an `await` keyword that does something
+similar, we just need to find a way of converting a `&PyAny` into a Rust future so we can use the
+`await` keyword on it.
```

Comment:
I feel the whole paragraph a little bit redundant. The key point is we need `pyo3_asyncio::into_future` to make the Python coroutine `await`able in Rust, so we might want to note that earlier than pointing out that both languages have the same keyword `await`.

Reply:
Yeah, I see what you mean. How about condensing it down to this?

> Normally in Python, that something special is the `await` keyword, but in order to await this coroutine in Rust, we first need to convert it into a `Future`. 

It could probably be joined up with line 34 too.